runtime/stdlib/random.go

```diff
@@ -195,3 +205,207 @@ func NewRevertibleRandomFunction(generator RandomGenerator) StandardLibraryValue
+// Although the final output is a `uint64`, it can be safely
+// casted into the desired output type because the extra bytes are guaranteed
+// to be zeros.
+func getUint64RandomNumber(
+	generator RandomGenerator,
+	ty sema.Type,
+	moduloArg interpreter.Value,
+) uint64 {
+
+	// buffer to get random bytes from the generator
+	// 8 is the size of the largest type supported
+	const bufferSize = 8
+	var buffer [bufferSize]byte
+
+	// case where no modulo argument was provided
+	if moduloArg == nil {
+		numericType, ok := ty.(*sema.NumericType)
+		if !ok {
+			// checker should prevent this
+			panic(errors.NewUnreachableError())
+		}
+		bytes := numericType.ByteSize()
+		getRandomBytes(buffer[bufferSize-bytes:], generator)
+		return binary.BigEndian.Uint64(buffer[:])
+	}
+
+	var ok bool
+	var modulo uint64
+
+	switch ty {
+	case sema.UInt8Type:
+		var moduloVal interpreter.UInt8Value
+		moduloVal, ok = moduloArg.(interpreter.UInt8Value)
+		modulo = uint64(moduloVal)
+	case sema.UInt16Type:
+		var moduloVal interpreter.UInt16Value
+		moduloVal, ok = moduloArg.(interpreter.UInt16Value)
+		modulo = uint64(moduloVal)
+	case sema.UInt32Type:
+		var moduloVal interpreter.UInt32Value
+		moduloVal, ok = moduloArg.(interpreter.UInt32Value)
+		modulo = uint64(moduloVal)
+	case sema.UInt64Type:
+		var moduloVal interpreter.UInt64Value
+		moduloVal, ok = moduloArg.(interpreter.UInt64Value)
+		modulo = uint64(moduloVal)
+	case sema.Word8Type:
+		var moduloVal interpreter.Word8Value
+		moduloVal, ok = moduloArg.(interpreter.Word8Value)
+		modulo = uint64(moduloVal)
+	case sema.Word16Type:
+		var moduloVal interpreter.Word16Value
+		moduloVal, ok = moduloArg.(interpreter.Word16Value)
+		modulo = uint64(moduloVal)
+	case sema.Word32Type:
+		var moduloVal interpreter.Word32Value
+		moduloVal, ok = moduloArg.(interpreter.Word32Value)
+		modulo = uint64(moduloVal)
+	case sema.Word64Type:
+		var moduloVal interpreter.Word64Value
+		moduloVal, ok = moduloArg.(interpreter.Word64Value)
+		modulo = uint64(moduloVal)
+	default:
+		// sanity check: shouldn't reach here
+		panic(errors.NewUnreachableError())
+	}
+
+	if !ok {
+		// checker should prevent this
+		panic(errors.NewUnreachableError())
+	}
+
+	// user error if modulo is zero
+	if modulo == 0 {
+		panic(ZeroModuloError)
+	}
+
+	// `max` is the maximum value that can be returned
+	max := modulo - 1
+	// get a bit mask (0b11..11) that covers all `max` bits,
+	// and count the byte size of `max`
+	mask := uint64(0)
+	bitSize := 0
+	for max&mask != max {
+		bitSize++
+		mask = (mask << 1) | 1
+	}
+	byteSize := (bitSize + 7) >> 3
+
+	// Generate a number less or equal than `max`.
+	// use the reject-sample method to avoid the modulo bias.
+	// the function isn't constant-time in this case and may take longer than computing
+	// a modular reduction.
+	// However, sampling exactly the size of `max` in bits makes the loop return fast:
+	// loop returns after (k) iterations with a probability of at most 1-(1/2)^k.
```

Comment:
I have problems following your argument of 
```
loop returns after (k) iterations with a probability of at most 1-(1/2)^k.
```
There is probably some detail that I am overlooking. So I'll try to explain my thoughts and please correct me, where I am mistaken:
* To the best of my understanding, we are interested in the probability bound of the _worst case_. Specifically, the _lower bound_. Hence, I would expect a statement like
   ``` 
   The probability of the loop running for k iterations is _at most_ (1/2)^k
   ``` 
   or
   ``` 
   The probability of the loop taking k iterations or less is _at least_ 1 - (1/2)^k
   ```

Reply:
I agree my sentence isn't clear, I'll update